### PR TITLE
trainer: clear MPS graph cache via torch_empty_cache_steps

### LINKED ISSUE #33717: Trainer class causes massive memory leak when using mps
### System Info

- `transformers` version: 4.44.2
- Platform: macOS-14.4-arm64-arm-64bit
- Python version: 3.12.2
- Huggingface_hub version: 0.24.5
- Safetensors version: 0.4.3
- Accelerate version: 0.33.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.4.1 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from datasets import load_dataset
from transformers import Trainer, TrainingArguments, LlamaConfig, LlamaForCausalLM, LlamaTokenizer, DataCollatorForLanguageModeling

dataset = load_dataset("roneneldan/TinyStories")
config = {
            "vocab_size": 32000,
            "hidden_size": 128,
            "num_hidden_layers": 4,
            "num_attention_heads": 8,
            "intermediate_size": 256,
            "hidden_act": "silu",
            "max_position_embeddings": 51,
            "initializer_range": 0.02,
            "rms_norm_eps": 1e-6,
        }

model_config = LlamaConfig(**config)
model = LlamaForCausalLM(model_config)
tokenizer = LlamaTokenizer.from_pretrained("meta-llama/Llama-2-7b")
tokenizer.pad_token_id = tokenizer.eos_token_id

dataset["train"] = dataset["train"].select(range(100000))
dataset["validation"] = dataset["validation"].select(range(1000))
def tokenize_function(examples):
    return tokenizer(examples["text"], max_length=4096)
tokenized_datasets = dataset.map(tokenize_function, batched=True)

data_collator = DataCollatorForLanguageModeling(tokenizer, mlm=False)

args = TrainingArguments(
    output_dir=f"llama-tiny-stories",
    per_device_train_batch_size=4,
    per_device_eval_batch_size=16,
    evaluation_strategy="steps",
    eval_steps=3000,
    logging_steps=100,
    gradient_accumulation_steps=1,
    num_train_epochs=6,
    weight_decay=1e-5,
    learning_rate=1e-4,
    save_steps=0,
    fp16=False,
    push_to_hub=False,
    use_mps_device=True,
    lr_scheduler_type="constant"
)  

trainer = Trainer(
    model=model,
    tokenizer=tokenizer,
    args=args,
    data_collator=data_collator,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["validation"],
)

trainer.train()
```

1. Open activity monitor
2. Run the snippet above (tested both in notebook and as a script).

Memory of python process will keep growing continuously and training will be super slow compared to when you set `use_mps_device=False` and add `use_cpu=True`.

### Expected behavior

1. I would expect the script to run with a constant memory consumption similar to when run as a cpu run.
2. I would expect the training to be faster that the cpu run.

### PR DESCRIPTION
**Pull request by anagnorisis2peripeteia**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=45818)

## Summary

Closes #33717

MPSGraph on Apple Silicon can retain compiled graphs for shape-varying training workloads. With variable-length batches, the compiled graph cache grows across training. This PR keeps the default behavior unchanged, but when users opt in to periodic device-cache clearing with `torch_empty_cache_steps`, it also clears the MPS graph cache in that same existing `training_step()` cache-clearing path.

This follows review feedback to use the existing `torch_empty_cache_steps` cache-management knob instead of clearing graph cache unconditionally.

## Why this matters

I used local benchmark scripts to compare MPS graph-cache policies on BERT fine-tuning with WikiText-2, batch size 8, max length 128, 200 iterations, and subprocess-isolated `psutil` RSS. The scripts are intentionally not included in the final PR diff, per review feedback, but the benchmark setup and results are kept here for reproducibility/context.

The main benchmark compared the default cache policy with clearing, freezing, and disabling the graph cache:

| Strategy | RSS delta | ms/iter | note |
|---|---:|---:|---|
| always (default cache policy) | +794 MB | 185 ms | unbounded growth |
| `clear_graph_cache()` every optimizer step | +338 MB | 217 ms | bounded, 1.17x |
| `freeze_graph_cache()` after warmup | +443 MB | 192 ms | bounded, 1.04x |
| never cache | +209 MB | 1412 ms | bounded, 7.63x |

A second local sweep varied `max_len` to show why a user-facing clear option is useful for long-context, high-shape-diversity workloads:

| max_len | unique padded lengths | freeze overhead | clear overhead | lower-overhead bounded option |
|---:|---:|---:|---:|---|
| 64 | 61 | 1.00x | 1.68x | freeze |
| 128 | 125 | 1.03x | 1.15x | freeze |
| 256 | 253 | 1.36x | 1.07x | clear |
| 512 | 428 | 5.88x | 1.51x | clear |
| 1024 | 436 | 5.36x | 1.42x | clear |

The crossover is around 200 unique padded lengths. The workloads that OOM in practice, especially longer-context training with more variable padding, are above that crossover. That is why this PR exposes the bounded-memory behavior through the existing opt-in `torch_empty_cache_steps` setting rather than changing default Trainer behavior.

<details>
<summary>Benchmark script outline</summary>

The local benchmark scripts, kept out of the final diff, used:

- `bert-base-uncased` masked-LM fine-tuning on WikiText-2 raw train text.
- `DataCollatorWithPadding`, so each batch is padded to the maximum sequence length in that batch.
- Fresh subprocesses per strategy, because macOS `ru_maxrss` is monotonic and current RSS needs process isolation for a fair comparison.
- `torch.mps.clear_graph_cache()` after each optimizer step for the clear strategy, matching `torch_empty_cache_steps=1` in this PR.
- `torch.mps.freeze_graph_cache()` after warmup for the freeze strategy.
- `torch.mps.set_graph_cache_policy("never")` for the no-cache strategy.

The core strategy loop was:

```python
for i in range(ITERS):
    optimizer.zero_grad()
    loss = model(**inputs_for_iteration(i)).loss
    loss.backward()
    optimizer.step()

    if strategy == "clear_per_iter":
        torch.mps.clear_graph_cache()
    elif strategy == "freeze_after_warmup" and i == freeze_at:
        torch.mps.freeze_graph_cache()

    torch.mps.synchronize()
```

The sweep repeated the same workload for `max_len` values `[64, 128, 256, 512, 1024]` and compared runtime overhead against the default `always` cache policy.

</details>

## Change

- Add a guarded `torch.mps.clear_graph_cache()` call beside the existing `clear_device_cache()` call in `training_step()`.
- Gate it behind:
  - `self.args.torch_empty_cache_steps is not None`
  - the same step interval check used by the existing cache-clearing path
  - `torch.backends.mps.is_available()`
  - `hasattr(torch.mps, "clear_graph_cache")`
- Remove benchmark scripts and MPS-only tests from the final code diff.

With `torch_empty_cache_steps=1`, MPS graph cache is cleared whenever the existing per-step device-cache clear runs. With larger values, it follows that interval. If unset, behavior is unchanged.

The `hasattr` guard keeps this a no-op on PyTorch versions that do not expose `torch.mps.clear_graph_cache()`.

## Validation

- `make style`
- `make check-code-quality`
- `make check-repo`
- `python -m pytest tests/trainer/test_trainer.py::TrainerIntegrationTest::test_double_train_wrap_once tests/trainer/test_trainer.py::TrainerIntegrationTest::test_training_arguments_are_left_untouched tests/trainer/test_training_args.py::TestTrainingArguments::test_torch_empty_cache_steps_requirements -q`
- `utils/tests_fetcher.py` was run; because `trainer.py` is central, it selected a broad CI slice. Local targeted Trainer checks above passed.

AI assistance was used to action reviewer feedback and run validation; the final code diff was kept to reviewer-requested scope.

## References

- pytorch/pytorch#154329 and pytorch/pytorch#164299 - upstream MPS memory-growth reports involving graph/cache behavior
- pytorch/pytorch#181485 - existing PyTorch MPS policy: `torch.mps.empty_cache()` clears `MPSGraphCache`
- #33717 - Trainer MPS memory growth report

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @sunmarc maybe?

**Comment by Tokarak** [comment]
Ah, this looks a lot like the bug I experienced in #45517

**Comment by anagnorisis2peripeteia** [comment]
> Ah, this looks a lot like the bug I experienced in #45517

Yes MPSGraphCache is the source of the memory swelling. Effectively unbounded growth with variable sized inputs.

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -1246,3 +1247,41 @@ def test_double_train_wrap_once(self):
         trainer.train()
         model_wrapped_after = trainer.model_wrapped
         self.assertIs(model_wrapped_before, model_wrapped_after, "should be not wrapped twice")
+
+
+class TrainerMPSGraphCacheTest(TestCasePlus):
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py]: we don't test mps backend on our ci, so let's not add this

File: benchmark/benches/bench_trainer_mps_graph_cache.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
+#!/usr/bin/env python3
+"""Benchmark: MPS graph cache policy impact on BERT fine-tuning (WikiText-2).
+
+With DataCollatorWithPadding each batch is padded to max(lengths_in_batch).
+The shape key is a single sequence-length dimension bounded by max_length.
+Unlike GNN batching, shape space is 1-D and finite -- but with long-context
```
Comment (SunMarc) [inline benchmark/benches/bench_trainer_mps_graph_cache.py]: remove the benchs please

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -1772,6 +1772,8 @@ def _run_epoch(
                             self.lr_scheduler.step()
 
                     model.zero_grad()
+                    if torch.backends.mps.is_available() and hasattr(torch.mps, "clear_graph_cache"):
+                        torch.mps.clear_graph_cache()
```
Comment (SunMarc) [inline src/transformers/trainer.py]: maybe we can put that with `clear_device_cache` ? so that instead of it being always triggered, the user will have to specify `torch_empty_cache_steps`

**Review by SunMarc — COMMENTED** [review]
Left a comment ! Thanks

**Comment by ksasso1028** [comment]
what are we waiting on :)

File: benchmark/benches/bench_trainer_mps_graph_cache.py (hunk 2 above)
Comment (anagnorisis2peripeteia) [reply benchmark/benches/bench_trainer_mps_graph_cache.py]: Removed both benchmark files from the PR. The current diff no longer includes anything under `benchmark/benches/`.

File: tests/trainer/test_trainer.py (hunk 1 above)
Comment (anagnorisis2peripeteia) [reply tests/trainer/test_trainer.py]: Removed these tests. The pushed diff no longer changes `tests/trainer/test_trainer.py`, so we are not adding MPS-specific coverage to CI.

File: src/transformers/trainer.py (hunk 3 above)
Comment (anagnorisis2peripeteia) [reply src/transformers/trainer.py]: Done. The graph-cache clear is now gated by `torch_empty_cache_steps` plus the MPS/API availability checks, so default behavior is unchanged and users opt in through the existing cache-clearing setting.

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -1772,7 +1772,12 @@ def _run_epoch(
                             self.lr_scheduler.step()
 
                     model.zero_grad()
-                    if torch.backends.mps.is_available() and hasattr(torch.mps, "clear_graph_cache"):
+                    if (
+                        self.args.torch_empty_cache_steps is not None
+                        and self.state.global_step % self.args.torch_empty_cache_steps == 0
+                        and torch.backends.mps.is_available()
+                        and hasattr(torch.mps, "clear_graph_cache")
+                    ):
```
Comment (SunMarc) [inline src/transformers/trainer.py]: i mean we have the following in training_step():

```
            if (
                self.args.torch_empty_cache_steps is not None
                and self.state.global_step % self.args.torch_empty_cache_steps == 0
            ):
                clear_device_cache()
```
just add the code there

Comment (anagnorisis2peripeteia) [reply src/transformers/trainer.py]: Done, moved it into the existing `training_step()` `torch_empty_cache_steps` block, immediately after `clear_device_cache()`, with the same MPS availability and `hasattr(torch.mps, "clear_graph_cache")` guards.

**Review by SunMarc — APPROVED** [review]
